src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -902,6 +933,86 @@ namespace cryptonote
   }
   //------------------------------------------------------------------------------------------------------------------------
   template<class t_core>
+  int t_cryptonote_protocol_handler<t_core>::handle_notify_tx_pool_inv(int command, NOTIFY_TX_POOL_INV::request& arg, cryptonote_connection_context& context)
+  {
+    MLOG_P2P_MESSAGE("Received NOTIFY_TX_POOL_INV (" << arg.t.size() << " txes)");
+
+    // Create a list to hold transaction hashes missing in our local tx pool.
+    std::vector<crypto::hash> missing_tx_hashes;
+
+    // Iterate over each advertised transaction hash and check our pool and our requested tracker.
+    for (const auto &tx_hash : arg.t)
+    {
+      if (!context.can_process_additional_request())
+      {
+        LOG_ERROR_CCONTEXT("Too many in-flight requests, cannot accept");
+        return 1;
+      }
+      // If we have the tx in our pool, also remove it from request manager and skip.
+      if (m_core.pool_has_tx(tx_hash))
+      {
+        m_request_manager.remove_announcement(tx_hash);
+        continue;
+      }
+      bool need_request = m_request_manager.add_announcement(tx_hash, context.m_connection_id);
+      context.add_announcement(tx_hash);
+      if (need_request) {
+        context.add_requested_from_peer();
+        missing_tx_hashes.push_back(tx_hash);
+        m_request_manager.add_request(tx_hash, context.m_connection_id);
+      }
+    }
+
+    // If there are missing transactions, send a RequestTxPoolTxs message.
+    if (!missing_tx_hashes.empty())
+    {
+      NOTIFY_REQUEST_TX_POOL_TXS::request req;
+      req.t = std::move(missing_tx_hashes);
+      post_notify<NOTIFY_REQUEST_TX_POOL_TXS>(req, context);
+      MLOG_P2P_MESSAGE("Requested " << req.t.size() << " missing transactions via RequestTxPoolTxs");
+    }
+    else
+    {
+      MLOG_P2P_MESSAGE("All advertised transactions are already in the pool or were requested recently");
+    }
+
+    return 1;
+  }
+  //------------------------------------------------------------------------------------------------------------------------
+  template<class t_core>
+  int t_cryptonote_protocol_handler<t_core>::handle_request_tx_pool_txs(int command, NOTIFY_REQUEST_TX_POOL_TXS::request& arg, cryptonote_connection_context& context)
+  {
+    MLOG_P2P_MESSAGE("Received NOTIFY_REQUEST_TX_POOL_TXS (" << arg.t.size() << " txes)");
+
+    std::vector<blobdata> txs;
+
+    // Iterate over requested txin hashes
+    for (const auto &tx_hash : arg.t)
+    {
+      context.add_requested_from_me();
+      // Attempt to get the transaction blob from the mempool;
+      cryptonote::blobdata tx_blob;
+      if (m_core.get_pool_transaction(tx_hash, tx_blob, cryptonote::relay_category::broadcasted))
+      {
+        context.add_sent();
+        txs.push_back(std::move(tx_blob));
```

Comment:
Should return with an error as soon as `txs` vector gets bigger than the max packet size (100 MB?). This is in addition to the duplicate hash checking.

Reply:
Read the comment above. This array can grow to over 9 GB if an attacker sends a specially crafted request.